File: lib/js/arg.js
'use strict';

let Sys = require("./sys.js");
let Caml = require("./caml.js");
let List = require("./list.js");
let $$Array = require("./array.js");
let Buffer = require("./buffer.js");
let $$String = require("./string.js");
let Caml_obj = require("./caml_obj.js");
let Caml_array = require("./caml_array.js");
let Pervasives = require("./pervasives.js");
let Caml_format = require("./caml_format.js");
let Caml_string = require("./caml_string.js");
let Caml_exceptions = require("./caml_exceptions.js");
let Caml_js_exceptions = require("./caml_js_exceptions.js");

let Bad = /* @__PURE__ */Caml_exceptions.create("Arg.Bad");

let Help = /* @__PURE__ */Caml_exceptions.create("Arg.Help");

let Stop = /* @__PURE__ */Caml_exceptions.create("Arg.Stop");

function assoc3(x, _l) {
  while (true) {
    let l = _l;
    if (l) {
      let match = l.hd;
      if (Caml_obj.equal(match[0], x)) {
        return match[1];
      }
      _l = l.tl;
      continue;
    }
    throw Caml_js_exceptions.internalMakeExn("Not_found");
  };
}

function split(s) {
  let i = $$String.index(s, /* '=' */61);
  let len = s.length;
  return [
    $$String.sub(s, 0, i),
    $$String.sub(s, i + 1 | 0, len - (i + 1 | 0) | 0)
  ];
}

function make_symlist(prefix, sep, suffix, l) {
  if (l) {
    return List.fold_left((x, y) => x + (sep + y), prefix + l.hd, l.tl) + suffix;
  } else {
    return "<none>";
  }
}

function help_action() {
  throw Caml_js_exceptions.internalFromExtension({
    RE_EXN_ID: Stop,
    _1: {
      TAG: "Unknown",
      _0: "-help"
    }
  });
}

function add_help(speclist) {
  let add1;
  try {
    assoc3("-help", speclist);
    add1 = /* [] */0;
  } catch (raw_exn) {
    let exn = Caml_js_exceptions.internalAnyToExn(raw_exn);
    if (exn.RE_EXN_ID === "Not_found") {
      add1 = {
        hd: [
          "-help",
          {
            TAG: "Unit",
            _0: help_action
          },
          " Display this list of options"
        ],
        tl: /* [] */0
      };
    } else {
      throw exn;
    }
  }
  let add2;
  try {
    assoc3("--help", speclist);
    add2 = /* [] */0;
  } catch (raw_exn$1) {
    let exn$1 = Caml_js_exceptions.internalAnyToExn(raw_exn$1);
    if (exn$1.RE_EXN_ID === "Not_found") {
      add2 = {
        hd: [
          "--help",
          {
            TAG: "Unit",
            _0: help_action
          },
          " Display this list of options"
        ],
        tl: /* [] */0
      };
    } else {
      throw exn$1;
    }
  }
  return Pervasives.$at(speclist, Pervasives.$at(add1, add2));
}

function usage_b(buf, speclist, errmsg) {
  Buffer.add_string(buf, errmsg + "\n");
  List.iter(x => {
    let doc = x[2];
    if (doc.length === 0) {
      return;
    }
    let spec = x[1];
    let key = x[0];
    if (spec.TAG !== "Symbol") {
      return Buffer.add_string(buf, "  " + key + " " + doc + "\n");
    }
    let sym = make_symlist("{", "|", "}", spec._0);
    Buffer.add_string(buf, "  " + key + " " + sym + doc + "\n");
  }, add_help(speclist));
}

function usage_string(speclist, errmsg) {
  let b = Buffer.create(200);
  usage_b(b, speclist, errmsg);
  return Buffer.contents(b);
}

function usage(speclist, errmsg) {
  console.log(usage_string(speclist, errmsg));
}

let current = {
  contents: 0
};

function bool_of_string_opt(x) {
  try {
    return Pervasives.bool_of_string(x);
  } catch (raw_exn) {
    let exn = Caml_js_exceptions.internalAnyToExn(raw_exn);
    if (exn.RE_EXN_ID === "Invalid_argument") {
      return;
    }
    throw exn;
  }
}

function int_of_string_opt(x) {
  try {
    return Caml_format.int_of_string(x);
  } catch (raw_exn) {
    let exn = Caml_js_exceptions.internalAnyToExn(raw_exn);
    if (exn.RE_EXN_ID === "Failure") {
      return;
    }
    throw exn;
  }
}

function float_of_string_opt(x) {
  try {
    return Caml_format.float_of_string(x);
  } catch (raw_exn) {
    let exn = Caml_js_exceptions.internalAnyToExn(raw_exn);
    if (exn.RE_EXN_ID === "Failure") {
      return;
    }
    throw exn;
  }
}

function parse_and_expand_argv_dynamic_aux(allow_expand, current, argv, speclist, anonfun, errmsg) {
  let initpos = current.contents;
  let convert_error = error => {
    let b = Buffer.create(200);
    let progname = initpos < argv.contents.length ? Caml_array.get(argv.contents, initpos) : "(?)";
    switch (error.TAG) {
      case "Unknown" :
        let s = error._0;
        switch (s) {
          case "--help" :
          case "-help" :
            break;
          default:
            Buffer.add_string(b, progname + ": unknown option '" + s + "'.\n");
        }
        break;
      case "Wrong" :
        Buffer.add_string(b, progname + ": wrong argument '" + error._1 + "'; option '" + error._0 + "' expects " + error._2 + ".\n");
        break;
      case "Missing" :
        Buffer.add_string(b, progname + ": option '" + error._0 + "' needs an argument.\n");
        break;
      case "Message" :
        Buffer.add_string(b, progname + ": " + error._0 + ".\n");
        break;
    }
    usage_b(b, speclist.contents, errmsg);
    if (Caml_obj.equal(error, {
        TAG: "Unknown",
        _0: "-help"
      }) || Caml_obj.equal(error, {
        TAG: "Unknown",
        _0: "--help"
      })) {
      return Caml_js_exceptions.internalFromExtension({
        RE_EXN_ID: Help,
        _1: Buffer.contents(b)
      });
    } else {
      return Caml_js_exceptions.internalFromExtension({
        RE_EXN_ID: Bad,
        _1: Buffer.contents(b)
      });
    }
  };
  current.contents = current.contents + 1 | 0;
  while (current.contents < argv.contents.length) {
    try {
      let s = Caml_array.get(argv.contents, current.contents);
      if (s.length >= 1 && Caml_string.get(s, 0) === /* '-' */45) {
        let match;
        try {
          match = [
            assoc3(s, speclist.contents),
            undefined
          ];
        } catch (raw_exn) {
          let exn = Caml_js_exceptions.internalAnyToExn(raw_exn);
          if (exn.RE_EXN_ID === "Not_found") {
            try {
              let match$1 = split(s);
              match = [
                assoc3(match$1[0], speclist.contents),
                match$1[1]
              ];
            } catch (raw_exn$1) {
              let exn$1 = Caml_js_exceptions.internalAnyToExn(raw_exn$1);
              if (exn$1.RE_EXN_ID === "Not_found") {
                throw Caml_js_exceptions.internalFromExtension({
                  RE_EXN_ID: Stop,
                  _1: {
                    TAG: "Unknown",
                    _0: s
                  }
                });
              }
              throw exn$1;
            }
          } else {
            throw exn;
          }
        }
        let follow = match[1];
        let no_arg = () => {
          if (follow === undefined) {
            return;
          }
          throw Caml_js_exceptions.internalFromExtension({
            RE_EXN_ID: Stop,
            _1: {
              TAG: "Wrong",
              _0: s,
              _1: follow,
              _2: "no argument"
            }
          });
        };
        let get_arg = () => {
          if (follow !== undefined) {
            return follow;
          }
          if ((current.contents + 1 | 0) < argv.contents.length) {
            return Caml_array.get(argv.contents, current.contents + 1 | 0);
          }
          throw Caml_js_exceptions.internalFromExtension({
            RE_EXN_ID: Stop,
            _1: {
              TAG: "Missing",
              _0: s
            }
          });
        };
        let consume_arg = () => {
          if (follow !== undefined) {
            return;
          } else {
            current.contents = current.contents + 1 | 0;
            return;
          }
        };
        let treat_action = f => {
          switch (f.TAG) {
            case "Unit" :
              return f._0();
            case "Bool" :
              let arg = get_arg();
              let s$1 = bool_of_string_opt(arg);
              if (s$1 !== undefined) {
                f._0(s$1);
              } else {
                throw Caml_js_exceptions.internalFromExtension({
                  RE_EXN_ID: Stop,
                  _1: {
                    TAG: "Wrong",
                    _0: s,
                    _1: arg,
                    _2: "a boolean"
                  }
                });
              }
              return consume_arg();
            case "Set" :
              no_arg();
              f._0.contents = true;
              return;
            case "Clear" :
              no_arg();
              f._0.contents = false;
              return;
            case "String" :
              let arg$1 = get_arg();
              f._0(arg$1);
              return consume_arg();
            case "Set_string" :
              f._0.contents = get_arg();
              return consume_arg();
            case "Int" :
              let arg$2 = get_arg();
              let x = int_of_string_opt(arg$2);
              if (x !== undefined) {
                f._0(x);
              } else {
                throw Caml_js_exceptions.internalFromExtension({
                  RE_EXN_ID: Stop,
                  _1: {
                    TAG: "Wrong",
                    _0: s,
                    _1: arg$2,
                    _2: "an integer"
                  }
                });
              }
              return consume_arg();
            case "Set_int" :
              let arg$3 = get_arg();
              let x$1 = int_of_string_opt(arg$3);
              if (x$1 !== undefined) {
                f._0.contents = x$1;
              } else {
                throw Caml_js_exceptions.internalFromExtension({
                  RE_EXN_ID: Stop,
                  _1: {
                    TAG: "Wrong",
                    _0: s,
                    _1: arg$3,
                    _2: "an integer"
                  }
                });
              }
              return consume_arg();
            case "Float" :
              let arg$4 = get_arg();
              let x$2 = float_of_string_opt(arg$4);
              if (x$2 !== undefined) {
                f._0(x$2);
              } else {
                throw Caml_js_exceptions.internalFromExtension({
                  RE_EXN_ID: Stop,
                  _1: {
                    TAG: "Wrong",
                    _0: s,
                    _1: arg$4,
                    _2: "a float"
                  }
                });
              }
              return consume_arg();
            case "Set_float" :
              let arg$5 = get_arg();
              let x$3 = float_of_string_opt(arg$5);
              if (x$3 !== undefined) {
                f._0.contents = x$3;
              } else {
                throw Caml_js_exceptions.internalFromExtension({
                  RE_EXN_ID: Stop,
                  _1: {
                    TAG: "Wrong",
                    _0: s,
                    _1: arg$5,
                    _2: "a float"
                  }
                });
              }
              return consume_arg();
            case "Tuple" :
              return List.iter(treat_action, f._0);
            case "Symbol" :
              let symb = f._0;
              let arg$6 = get_arg();
              if (List.mem(arg$6, symb)) {
                f._1(arg$6);
                return consume_arg();
              }
              throw Caml_js_exceptions.internalFromExtension({
                RE_EXN_ID: Stop,
                _1: {
                  TAG: "Wrong",
                  _0: s,
                  _1: arg$6,
                  _2: "one of: " + make_symlist("", " ", "", symb)
                }
              });
            case "Rest" :
              let f$1 = f._0;
              while (current.contents < (argv.contents.length - 1 | 0)) {
                f$1(Caml_array.get(argv.contents, current.contents + 1 | 0));
                consume_arg();
              };
              return;
            case "Expand" :
              if (!allow_expand) {
                throw Caml_js_exceptions.internalFromExtension({
                  RE_EXN_ID: "Invalid_argument",
                  _1: "Arg.Expand is is only allowed with Arg.parse_and_expand_argv_dynamic"
                });
              }
              let arg$7 = get_arg();
              let newarg = f._0(arg$7);
              consume_arg();
              let before = $$Array.sub(argv.contents, 0, current.contents + 1 | 0);
              let after = $$Array.sub(argv.contents, current.contents + 1 | 0, (argv.contents.length - current.contents | 0) - 1 | 0);
              argv.contents = Caml_array.concat({
                hd: before,
                tl: {
                  hd: newarg,
                  tl: {
                    hd: after,
                    tl: /* [] */0
                  }
                }
              });
              return;
          }
        };
        treat_action(match[0]);
      } else {
        anonfun(s);
      }
    } catch (raw_m) {
      let m = Caml_js_exceptions.internalAnyToExn(raw_m);
      if (m.RE_EXN_ID === Bad) {
        throw convert_error({
          TAG: "Message",
          _0: m._1
        });
      }
      if (m.RE_EXN_ID === Stop) {
        throw convert_error(m._1);
      }
      throw m;
    }
    current.contents = current.contents + 1 | 0;
  };
}

function parse_and_expand_argv_dynamic(current, argv, speclist, anonfun, errmsg) {
  parse_and_expand_argv_dynamic_aux(true, current, argv, speclist, anonfun, errmsg);
}

function parse_argv_dynamic(currentOpt, argv, speclist, anonfun, errmsg) {
  let current$1 = currentOpt !== undefined ? currentOpt : current;
  parse_and_expand_argv_dynamic_aux(false, current$1, {
    contents: argv
  }, speclist, anonfun, errmsg);
}

function parse_argv(currentOpt, argv, speclist, anonfun, errmsg) {
  let current$1 = currentOpt !== undefined ? currentOpt : current;
  parse_argv_dynamic(current$1, argv, {
    contents: speclist
  }, anonfun, errmsg);
}

function parse(l, f, msg) {
  try {
    return parse_argv(undefined, Sys.argv, l, f, msg);
  } catch (raw_msg) {
    let msg$1 = Caml_js_exceptions.internalAnyToExn(raw_msg);
    if (msg$1.RE_EXN_ID === Bad) {
      console.log(msg$1._1);
      return Pervasives.exit(2);
    }
    if (msg$1.RE_EXN_ID === Help) {
      console.log(msg$1._1);
      return Pervasives.exit(0);
    }
    throw msg$1;
  }
}

function parse_dynamic(l, f, msg) {
  try {
    return parse_argv_dynamic(undefined, Sys.argv, l, f, msg);
  } catch (raw_msg) {
    let msg$1 = Caml_js_exceptions.internalAnyToExn(raw_msg);
    if (msg$1.RE_EXN_ID === Bad) {
      console.log(msg$1._1);
      return Pervasives.exit(2);
    }
    if (msg$1.RE_EXN_ID === Help) {
      console.log(msg$1._1);
      return Pervasives.exit(0);
    }
    throw msg$1;
  }
}

function parse_expand(l, f, msg) {
  try {
    let argv = {
      contents: Sys.argv
    };
    let spec = {
      contents: l
    };
    let current$1 = {
      contents: current.contents
    };
    return parse_and_expand_argv_dynamic(current$1, argv, spec, f, msg);
  } catch (raw_msg) {
    let msg$1 = Caml_js_exceptions.internalAnyToExn(raw_msg);
    if (msg$1.RE_EXN_ID === Bad) {
      console.log(msg$1._1);
      return Pervasives.exit(2);
    }
    if (msg$1.RE_EXN_ID === Help) {
      console.log(msg$1._1);
      return Pervasives.exit(0);
    }
    throw msg$1;
  }
}

function second_word(s) {
  let len = s.length;
  let loop = _n => {
    while (true) {
      let n = _n;
      if (n >= len) {
        return len;
      }
      if (Caml_string.get(s, n) !== /* ' ' */32) {
        return n;
      }
      _n = n + 1 | 0;
      continue;
    };
  };
  let n;
  try {
    n = $$String.index(s, /* '\t' */9);
  } catch (raw_exn) {
    let exn = Caml_js_exceptions.internalAnyToExn(raw_exn);
    if (exn.RE_EXN_ID === "Not_found") {
      let exit = 0;
      let n$1;
      try {
        n$1 = $$String.index(s, /* ' ' */32);
        exit = 2;
      } catch (raw_exn$1) {
        let exn$1 = Caml_js_exceptions.internalAnyToExn(raw_exn$1);
        if (exn$1.RE_EXN_ID === "Not_found") {
          return len;
        }
        throw exn$1;
      }
      if (exit === 2) {
        return loop(n$1 + 1 | 0);
      }
      
    } else {
      throw exn;
    }
  }
  return loop(n + 1 | 0);
}

function max_arg_len(cur, param) {
  let kwd = param[0];
  if (param[1].TAG === "Symbol") {
    return Caml.int_max(cur, kwd.length);
  } else {
    return Caml.int_max(cur, kwd.length + second_word(param[2]) | 0);
  }
}

function replace_leading_tab(s) {
  let seen = {
    contents: false
  };
  return $$String.map(c => {
    if (c !== 9 || seen.contents) {
      return c;
    } else {
      seen.contents = true;
      return /* ' ' */32;
    }
  }, s);
}

function align(limitOpt, speclist) {
  let limit = limitOpt !== undefined ? limitOpt : Pervasives.max_int;
  let completed = add_help(speclist);
  let len = List.fold_left(max_arg_len, 0, completed);
  let len$1 = len < limit ? len : limit;
  return List.map(x => {
    let spec = x[1];
    let kwd = x[0];
    if (x[2] === "") {
      return x;
    }
    if (spec.TAG === "Symbol") {
      let msg = x[2];
      let cutcol = second_word(msg);
      let spaces = " ".repeat(Caml.int_max(0, len$1 - cutcol | 0) + 3 | 0);
      return [
        kwd,
        spec,
        "\n" + (spaces + replace_leading_tab(msg))
      ];
    }
    let msg$1 = x[2];
    let cutcol$1 = second_word(msg$1);
    let kwd_len = kwd.length;
    let diff = (len$1 - kwd_len | 0) - cutcol$1 | 0;
    if (diff <= 0) {
      return [
        kwd,
        spec,
        replace_leading_tab(msg$1)
      ];
    }
    let spaces$1 = " ".repeat(diff);
    let prefix = $$String.sub(replace_leading_tab(msg$1), 0, cutcol$1);
    let suffix = $$String.sub(msg$1, cutcol$1, msg$1.length - cutcol$1 | 0);
    return [
      kwd,
      spec,
      prefix + (spaces$1 + suffix)
    ];
  }, completed);
}

exports.parse = parse;
exports.parse_dynamic = parse_dynamic;
exports.parse_argv = parse_argv;
exports.parse_argv_dynamic = parse_argv_dynamic;
exports.parse_and_expand_argv_dynamic = parse_and_expand_argv_dynamic;
exports.parse_expand = parse_expand;
exports.Help = Help;
exports.Bad = Bad;
exports.usage = usage;
exports.usage_string = usage_string;
exports.align = align;
exports.current = current;
/* No side effect */

File: jscomp/stdlib-406/arg.ml
(**************************************************************************)
(*                                                                        *)
(*                                 OCaml                                  *)
(*                                                                        *)
(*              Damien Doligez, projet Para, INRIA Rocquencourt           *)
(*                                                                        *)
(*   Copyright 1996 Institut National de Recherche en Informatique et     *)
(*     en Automatique.                                                    *)
(*                                                                        *)
(*   All rights reserved.  This file is distributed under the terms of    *)
(*   the GNU Lesser General Public License version 2.1, with the          *)
(*   special exception on linking described in the file LICENSE.          *)
(*                                                                        *)
(**************************************************************************)

type key = string
type doc = string
type usage_msg = string
type anon_fun = (string -> unit)

type spec =
  | Unit of (unit -> unit)     (* Call the function with unit argument *)
  | Bool of (bool -> unit)     (* Call the function with a bool argument *)
  | Set of bool ref            (* Set the reference to true *)
  | Clear of bool ref          (* Set the reference to false *)
  | String of (string -> unit) (* Call the function with a string argument *)
  | Set_string of string ref   (* Set the reference to the string argument *)
  | Int of (int -> unit)       (* Call the function with an int argument *)
  | Set_int of int ref         (* Set the reference to the int argument *)
  | Float of (float -> unit)   (* Call the function with a float argument *)
  | Set_float of float ref     (* Set the reference to the float argument *)
  | Tuple of spec list         (* Take several arguments according to the
                                  spec list *)
  | Symbol of string list * (string -> unit)
                               (* Take one of the symbols as argument and
                                  call the function with the symbol. *)
  | Rest of (string -> unit)   (* Stop interpreting keywords and call the
                                  function with each remaining argument *)
  | Expand of (string -> string array) (* If the remaining arguments to process
                                          are of the form
                                          [["-foo"; "arg"] @ rest] where "foo" is
                                          registered as [Expand f], then the
                                          arguments [f "arg" @ rest] are
                                          processed. Only allowed in
                                          [parse_and_expand_argv_dynamic]. *)

exception Bad of string
exception Help of string

type error =
  | Unknown of string
  | Wrong of string * string * string  (* option, actual, expected *)
  | Missing of string
  | Message of string

exception Stop of error (* used internally *)



let rec assoc3 x l =
  match l with
  | [] -> raise Not_found
  | (y1, y2, _) :: _ when y1 = x -> y2
  | _ :: t -> assoc3 x t


let split s =
  let i = String.index s '=' in
  let len = String.length s in
  String.sub s 0 i, String.sub s (i+1) (len-(i+1))


let make_symlist prefix sep suffix l =
  match l with
  | [] -> "<none>"
  | h::t -> (List.fold_left (fun x y -> x ^ sep ^ y) (prefix ^ h) t) ^ suffix


let print_spec buf (key, spec, doc) =
  if String.length doc > 0 then
    match spec with
    | Symbol (l, _) ->
#if 0    
        bprintf buf "  %s %s%s\n" key (make_symlist "{" "|" "}" l) doc
#else
        let sym = make_symlist "{" "|" "}" l in 
        Buffer.add_string buf {j|  $(key) $(sym)$(doc)\n|j}
#end
    | _ ->
#if 0    
        bprintf buf "  %s %s\n" key doc
#else
        Buffer.add_string buf {j|  $(key) $(doc)\n|j}
#end


let help_action () = raise (Stop (Unknown "-help"))

let add_help speclist =
  let add1 =
    try ignore (assoc3 "-help" speclist); []
    with Not_found ->
            ["-help", Unit help_action, " Display this list of options"]
  and add2 =
    try ignore (assoc3 "--help" speclist); []
    with Not_found ->
            ["--help", Unit help_action, " Display this list of options"]
  in
  speclist @ (add1 @ add2)


let usage_b buf speclist errmsg =
#if 0  
  bprintf buf "%s\n" errmsg;
#else
  Buffer.add_string buf {j|$(errmsg)\n|j};
#end
  List.iter (print_spec buf) (add_help speclist)


let usage_string speclist errmsg =
  let b = Buffer.create 200 in
  usage_b b speclist errmsg;
  Buffer.contents b


let usage speclist errmsg =
  Js.log (usage_string speclist errmsg)


let current = ref 0

let bool_of_string_opt x =
  try Some (bool_of_string x)
  with Invalid_argument _ -> None

let int_of_string_opt x =
  try Some (int_of_string x)
  with Failure _ -> None

let float_of_string_opt x =
  try Some (float_of_string x)
  with Failure _ -> None

let parse_and_expand_argv_dynamic_aux allow_expand current argv speclist anonfun errmsg =
  let initpos = !current in
  let convert_error error =
    (* convert an internal error to a Bad/Help exception
       *or* add the program name as a prefix and the usage message as a suffix
       to an user-raised Bad exception.
    *)
    let b = Buffer.create 200 in
    let progname = if initpos < (Array.length !argv) then !argv.(initpos) else "(?)" in
    begin match error with
      | Unknown "-help" -> ()
      | Unknown "--help" -> ()
      | Unknown s ->
#if 0      
          bprintf b "%s: unknown option '%s'.\n" progname s
#else
          Buffer.add_string b {j|$(progname): unknown option '$(s)'.\n|j} 
#end          
      | Missing s ->
#if 0      
          bprintf b "%s: option '%s' needs an argument.\n" progname s
#else
          Buffer.add_string b {j|$(progname): option '$(s)' needs an argument.\n|j}
#end          
      | Wrong (opt, arg, expected) ->
#if 0      
          bprintf b "%s: wrong argument '%s'; option '%s' expects %s.\n"
                  progname arg opt expected
#else
          Buffer.add_string b {j|$(progname): wrong argument '$(arg)'; option '$(opt)' expects $(expected).\n|j}
#end                  
      | Message s -> (* user error message *)
#if 0      
          bprintf b "%s: %s.\n" progname s
#else
          Buffer.add_string b {j|$(progname): $(s).\n|j}
#end          
    end;
    usage_b b !speclist errmsg;
    if error = Unknown "-help" || error = Unknown "--help"
    then Help (Buffer.contents b)
    else Bad (Buffer.contents b)
  in
  incr current;
  while !current < (Array.length !argv) do
    begin try
      let s = !argv.(!current) in
      if String.length s >= 1 && s.[0] = '-' then begin
        let action, follow =
          try assoc3 s !speclist, None
          with Not_found ->
          try
            let keyword, arg = split s in
            assoc3 keyword !speclist, Some arg
          with Not_found -> raise (Stop (Unknown s))
        in
        let no_arg () =
          match follow with
          | None -> ()
          | Some arg -> raise (Stop (Wrong (s, arg, "no argument"))) in
        let get_arg () =
          match follow with
          | None ->
              if !current + 1 < (Array.length !argv) then !argv.(!current + 1)
              else raise (Stop (Missing s))
          | Some arg -> arg
        in
        let consume_arg () =
          match follow with
          | None -> incr current
          | Some _ -> ()
        in
        let rec treat_action = function
        | Unit f -> f ();
        | Bool f ->
            let arg = get_arg () in
            begin match bool_of_string_opt arg with
            | None -> raise (Stop (Wrong (s, arg, "a boolean")))
            | Some s -> f s
            end;
            consume_arg ();
        | Set r -> no_arg (); r := true;
        | Clear r -> no_arg (); r := false;
        | String f ->
            let arg = get_arg () in
            f arg;
            consume_arg ();
        | Symbol (symb, f) ->
            let arg = get_arg () in
            if List.mem arg symb then begin
              f arg;
              consume_arg ();
            end else begin
              raise (Stop (Wrong (s, arg, "one of: "
                                          ^ (make_symlist "" " " "" symb))))
            end
        | Set_string r ->
            r := get_arg ();
            consume_arg ();
        | Int f ->
            let arg = get_arg () in
            begin match int_of_string_opt arg with
            | None -> raise (Stop (Wrong (s, arg, "an integer")))
            | Some x -> f x
            end;
            consume_arg ();
        | Set_int r ->
            let arg = get_arg () in
            begin match int_of_string_opt arg with
            | None -> raise (Stop (Wrong (s, arg, "an integer")))
            | Some x -> r := x
            end;
            consume_arg ();
        | Float f ->
            let arg = get_arg () in
            begin match float_of_string_opt arg with
            | None -> raise (Stop (Wrong (s, arg, "a float")))
            | Some x -> f x
            end;
            consume_arg ();
        | Set_float r ->
            let arg = get_arg () in
            begin match float_of_string_opt arg with
            | None -> raise (Stop (Wrong (s, arg, "a float")))
            | Some x -> r := x
            end;
            consume_arg ();
        | Tuple specs ->
            List.iter treat_action specs;
        | Rest f ->
            while !current < (Array.length !argv) - 1 do
              f !argv.(!current + 1);
              consume_arg ();
            done;
        | Expand f ->
            if not allow_expand then
              raise (Invalid_argument "Arg.Expand is is only allowed with Arg.parse_and_expand_argv_dynamic");
            let arg = get_arg () in
            let newarg = f arg in
            consume_arg ();
            let before = Array.sub !argv 0 (!current + 1)
            and after = Array.sub !argv (!current + 1) ((Array.length !argv) - !current - 1) in
            argv:= Array.concat [before;newarg;after];
        in
        treat_action action end
      else anonfun s
    with | Bad m -> raise (convert_error (Message m));
         | Stop e -> raise (convert_error e);
    end;
    incr current
  done

let parse_and_expand_argv_dynamic current argv speclist anonfun errmsg =
  parse_and_expand_argv_dynamic_aux true current argv speclist anonfun errmsg

let parse_argv_dynamic ?(current=current) argv speclist anonfun errmsg =
  parse_and_expand_argv_dynamic_aux false current (ref argv) speclist anonfun errmsg


let parse_argv ?(current=current) argv speclist anonfun errmsg =
  parse_argv_dynamic ~current:current argv (ref speclist) anonfun errmsg


let parse l f msg =
  try
    parse_argv Sys.argv l f msg
  with
  | Bad msg -> Js.log msg; exit 2
  | Help msg -> Js.log msg; exit 0


let parse_dynamic l f msg =
  try
    parse_argv_dynamic Sys.argv l f msg
  with
  | Bad msg -> Js.log msg; exit 2
  | Help msg -> Js.log msg; exit 0

let parse_expand l f msg =
  try
    let argv = ref Sys.argv in
    let spec = ref l in
    let current = ref (!current) in
    parse_and_expand_argv_dynamic current argv spec f msg
  with
  | Bad msg -> Js.log msg; exit 2
  | Help msg -> Js.log msg; exit 0


let second_word s =
  let len = String.length s in
  let rec loop n =
    if n >= len then len
    else if s.[n] = ' ' then loop (n+1)
    else n
  in
  match String.index s '\t' with
  | n -> loop (n+1)
  | exception Not_found ->
      begin match String.index s ' ' with
      | n -> loop (n+1)
      | exception Not_found -> len
      end


let max_arg_len cur (kwd, spec, doc) =
  match spec with
  | Symbol _ -> max cur (String.length kwd)
  | _ -> max cur (String.length kwd + second_word doc)


let replace_leading_tab s =
  let seen = ref false in
  String.map (function '\t' when not !seen -> seen := true; ' ' | c -> c) s

let add_padding len ksd =
  match ksd with
  | (_, _, "") ->
      (* Do not pad undocumented options, so that they still don't show up when
       * run through [usage] or [parse]. *)
      ksd
  | (kwd, (Symbol _ as spec), msg) ->
      let cutcol = second_word msg in
      let spaces = String.make ((max 0 (len - cutcol)) + 3) ' ' in
      (kwd, spec, "\n" ^ spaces ^ replace_leading_tab msg)
  | (kwd, spec, msg) ->
      let cutcol = second_word msg in
      let kwd_len = String.length kwd in
      let diff = len - kwd_len - cutcol in
      if diff <= 0 then
        (kwd, spec, replace_leading_tab msg)
      else
        let spaces = String.make diff ' ' in
        let prefix = String.sub (replace_leading_tab msg) 0 cutcol in
        let suffix = String.sub msg cutcol (String.length msg - cutcol) in
        (kwd, spec, prefix ^ spaces ^ suffix)


let align ?(limit=max_int) speclist =
  let completed = add_help speclist in
  let len = List.fold_left max_arg_len 0 completed in
  let len = min len limit in
  List.map (add_padding len) completed
#if 0
let trim_cr s =
  let len = String.length s in
  if len > 0 && String.get s (len - 1) = '\r' then
    String.sub s 0 (len - 1)
  else
    s

let read_aux trim sep file =
  let ic = open_in_bin file in
  let buf = Buffer.create 200 in
  let words = ref [] in
  let stash () =
    let word =  (Buffer.contents buf) in
    let word = if trim then trim_cr word else word in
    words := word :: !words;
    Buffer.clear buf
  in
  let rec read () =
    try
      let c = input_char ic in
      if c = sep then begin
        stash (); read ()
      end else begin
        Buffer.add_char buf c; read ()
      end
    with End_of_file ->
      if Buffer.length buf > 0 then
        stash () in
  read ();
  close_in ic;
  Array.of_list (List.rev !words)

let read_arg = read_aux true '\n'

let read_arg0 = read_aux false '\x00'

let write_aux sep file args =
  let oc = open_out_bin file in
  Array.iter (fun s -> fprintf oc "%s%c" s sep) args;
  close_out oc

let write_arg = write_aux '\n'

let write_arg0 = write_aux '\x00'
#end